runtime/commonMain/src/kotlinx/serialization/SchemaCache.kt

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2017-2019 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization
+
+private typealias DescriptorData<T> = MutableMap<DescriptorSchemaCache.Key<T>, T>
+
+/**
+ * A type-safe map for storing custom information (such as format schema), associated with [SerialDescriptor].
+ *
+ * This map is not thread safe. For thread safety and performance reasons, it is advised to pass it via stack (e.g. in function parameters)
+ */
+public class DescriptorSchemaCache {
```

Comment:
AFAIR, we've discussed some ideas about public cache which associates some specific information with descriptors that every serialization format could benefit from. It looks like a better solution instead of 'let 3rd party format authors write their own cache implementations again and again'.

I can agree that this class is generic and vague now, but we still have time to collect feedback about it. Since the library is still experimental, we can continue to change this class and adapt it for our purposes (next candidate, as you've correctly noticed, is protobuf). If experiments show that such format-specific caches can't be fit in a single abstraction, we can just deprecate and remove it.

Reply:
>AFAIR, we've discussed some ideas about public cache

Nope, about **just** a cache :)

We are again at the design decision "let's add something, maybe it's useful, maybe it's not, we can potentially remove it someday" which is not really a healthy way to do things, even though the library is still experimental.

We neither know if it has any use at all (at current point, at least), nor we will be able to know at some point. If someone uses this API, we will never know about it. Especially taking into account that there is at most dozen people who are going to write its own format. 

It's much better to adapt in in JSON/ProtoBuf, experiment (and benchmark!) different strategies and only then make the decision, not the vice versa.